### PR TITLE
Replace Split with Commutative Arrow. Continuing #1719

### PR DESCRIPTION
**Pull request by kailuowang**
per conversation with @diesalbla, this it the PR attempting to complete #1719

### CODE REVIEW

**Comment by edmundnoble** [comment]
My review is at the moment to use `locally` on line 42 of `CokleisliTests`.

File: core/src/main/scala/cats/data/Kleisli.scala [hunk 1]
```diff
@@ -91,11 +91,11 @@ private[data] sealed abstract class KleisliInstances extends KleisliInstances0 {
...
-    new KleisliArrow[F] { def F: Monad[F] = M }
+  implicit def catsDataCommutativeArrowForKleisli[F[_]](implicit M: CommutativeMonad[F]): CommutativeArrow[Kleisli[F, ?, ?]] =
+    new KleisliCommutativeArrow[F] {def F: CommutativeMonad[F] = M }
 
-  implicit val catsDataArrowForKleisliId: Arrow[Kleisli[Id, ?, ?]] =
-    catsDataArrowForKleisli[Id]
+  implicit val catsDataCommutativeArrowForKleisliId: CommutativeArrow[Kleisli[Id, ?, ?]] =
+    catsDataCommutativeArrowForKleisli[Id]
 
   implicit def catsDataMonadReaderForKleisliId[A]: MonadReader[Kleisli[Id, A, ?], A] =
```
Comment (edmundnoble) [inline core/src/main/scala/cats/data/Kleisli.scala]: This is seemingly a commutative monad now too, given `F[_]: Commutative`.

File: laws/src/main/scala/cats/laws/CommutativeCoflatMapLaws.scala [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Laws that must be obeyed by any `CommutativeCoflatMap`.
+ */
+trait CommutativeCoflatMapLaws[F[_]] extends CoflatMapLaws[F] {
+  implicit override def F: CommutativeCoflatMap[F]
+
+  def coflatmapCommutative[A, B, C](fa: F[A], fb: F[B], g: (F[A], F[B]) => C): IsEq[F[F[C]]] =
+    F.coflatMap(fa)(x => F.coflatMap(fb)(y => g(x, y))) <->
+    F.coflatMap(fb)(y => F.coflatMap(fa)(x => g(x, y)))
```
Comment (LukaJCB) [inline laws/src/main/scala/cats/laws/CommutativeCoflatMapLaws.scala]: Tiny nitpick, but should we name these `a` and `b` instead of `x` and `y`?

**Comment by kailuowang** [comment]
@edmundnoble you are right, both are `CommutativeMonad`, instances added.

**Comment by peterneyens** [comment]
Do we now if there are other `CommutativeCoflatMap` and/or `CommutativeComonad` instances than for `Id`? Otherwise I am not sure if these are worth adding ... ?

We can also make the `Arrow[Function1]` instance a `CommutativeArrow[Function1]`, right?

File: tests/src/test/scala/cats/tests/NonEmptyListTests.scala [hunk 3]
```diff
@@ -2,8 +2,9 @@ package cats
 package tests
 
 import cats.kernel.laws.{GroupLaws, OrderLaws}
+
 import cats.data.{NonEmptyList, NonEmptyVector}
-import cats.laws.discipline.{ComonadTests, MonadTests, NonEmptyTraverseTests, ReducibleTests, SemigroupKTests, SerializableTests}
+import cats.laws.discipline.{SemigroupKTests, MonadTests, SerializableTests, NonEmptyTraverseTests, ReducibleTests, ComonadTests}
```
Comment (peterneyens) [inline tests/src/test/scala/cats/tests/NonEmptyListTests.scala]: Maybe we can leave these alphabetically?

File: tests/src/test/scala/cats/tests/KleisliTests.scala [hunk 4]
```diff
@@ -238,9 +254,8 @@ class KleisliTests extends CatsSuite {
     MonadReader[Reader[Int, ?], Int]
     Monoid[Reader[Int, String]]
     MonoidK[λ[α => Reader[α, α]]]
-    Arrow[Reader[?, ?]]
+    CommutativeArrow[Reader[?, ?]]
```
Comment (peterneyens) [inline tests/src/test/scala/cats/tests/KleisliTests.scala]: We can leave `Arrow[Reader[?, ?]]` here (move it below `CommutativeArrow`).

**Comment by kailuowang** [comment]
@peterneyens I made function1 also commutative and correct the tests. 
The `CommutativeComonad` was introduced because `Reader` is a `CommutativeArrow` thanks to `Id` being a `CommutativeComonad`. Other than that I am not aware any other instances of `CommutativeComonad`, @diesalbla thought `NonEmptyList` is one but it turns out it isn't (or at least the law doesn't think so). I think there is value for `Reader` to be `CommutativeArrow`. We could remove `CommutativeComonad` and make a dedicated `CommutativeArrow` instance for `Reader`, not sure if that's a better solution though.

**Comment by peterneyens** [comment]
The `CommunativeComonad` is used in the `CommutativeArrow` instance of `Cokleisli`, right?
`Reader` has a `CommutativeArrow` because `Id` has a `CommutativeMonad`.

I was just thinking that if we can't come up with other instances, adding `CommunativeCoflatMap` and `CommutativeComonad` might not be worth it, given that it currently only gives us a `CommutativeArrow` for `Cokleisli[Id, ?, ?]` (also see #1364).

I think this PR is good to be merged :+1: (thanks @diesalbla and @kailuowang) ; I just wanted to hear other opinions if `CommutativeCoflatMap` and `CommutativeComonad` are useful enough to merit inclusion.

**Comment by edmundnoble** [comment]
Agreed with @peterneyens, I'd need at least one example of a commutative comonad/coflatmap before adding that but otherwise :+1: for merging.

**Comment by kailuowang** [comment]
@peterneyens you are right, it was `Cokleisli` that is benefiting from it not `Reader`, I agree with you and @edmundnoble , will remove `CommutativeComonad` and `CommutativeCoflatmap` for now

File: tests/src/test/scala/cats/tests/CokleisliTests.scala [hunk 5]
```diff
@@ -33,17 +33,20 @@ class CokleisliTests extends SlowCatsSuite {
...
 
   checkAll("Cokleisli[NonEmptyList, Int, Int]", MonoidKTests[λ[α => Cokleisli[NonEmptyList, α, α]]].monoidK[Int])
   checkAll("MonoidK[λ[α => Cokleisli[NonEmptyList, α, α]]]", SerializableTests.serializable(MonoidK[λ[α => Cokleisli[NonEmptyList, α, α]]]))
+  locally {
+    implicit def cokleisliIdEq[A, B](implicit A: Arbitrary[A], FB: Eq[B]): Eq[Cokleisli[Id, A, B]] =
+      Eq.by[Cokleisli[Id, A, B], A => B](_.run)
+
+    checkAll("Cokleisli[Id, Int, Int]", CommutativeArrowTests[Cokleisli[Id, ?, ?]].commutativeArrow[Int, Int, Int, Int, Int, Int])
+    checkAll("CommutativeArrow[Cokleisli[Id, ?, ?]]", SerializableTests.serializable(CommutativeArrow[Cokleisli[Id, ?, ?]]))
+  }
```
Comment (peterneyens) [inline tests/src/test/scala/cats/tests/CokleisliTests.scala]: I think we need to add a test for `Cokleisli[F, ?, ?]` as well (maybe the existing one using `NonEmptyList`)?

Nitpicking, can we put these `CommutativeArrow` tests not between the `MonoidK` and `SemigroupK` tests?

Comment (kailuowang) [inline tests/src/test/scala/cats/tests/CokleisliTests.scala]: addressed.

**Comment by diesalbla** [comment]
Regarding `CommutativeCoflatmap` and `CommutativeComonad`, the only reason I introduced these classes in the first case was to preserve the similarity between the `Arrow` (and relatives) instances for `Kleisli` and `Cokleisli`. I could not come up with too many instances for them.

**Comment by kailuowang** [comment]
merging with 2 sign-offs

**Pull request merged**